src/nodes.coffee

```diff
@@ -1296,6 +1314,115 @@ exports.LineComment = class LineComment extends Base
 
 #### Call
 
+exports.CSXElement = class CSXElement extends Base
+  constructor: ({@tagName, @attributes, @content}) ->
+    super()
+
+  children: ['tagName', 'attributes', 'content']
+
+  compileNode: (o) ->
+    @attributes.base.csx = yes
+    @content?.base.csx = yes
+    fragments = [@makeCode('<')]
+    fragments.push (tag = @tagName.compileToFragments(o, LEVEL_ACCESS))...
+    if @attributes.base instanceof Arr
+      for obj in @attributes.base.objects
+        attr = obj.base
+        attrProps = attr?.properties or []
+        # Catch invalid CSX attributes: <div {a:"b", props} {props} "value" />
+        if not (attr instanceof Obj or attr instanceof IdentifierLiteral) or (attr instanceof Obj and not attr.generated and (attrProps.length > 1 or not (attrProps[0] instanceof Splat)))
+          obj.error """
+            Unexpected token. Allowed CSX attributes are: id="val", src={source}, {props...} or attribute.
+          """
+        obj.base.csx = yes if obj.base instanceof Obj
+        fragments.push @makeCode ' '
+        fragments.push obj.compileToFragments(o, LEVEL_PAREN)...
+    if @content
+      fragments.push @makeCode('>')
+      fragments.push @content.compileNode(o, LEVEL_LIST)...
+      fragments.push [@makeCode('</'), tag..., @makeCode('>')]...
+    else
+      fragments.push @makeCode(' />')
+    fragments
+
+  ast: ->
+    # The location data spanning the opening element < ... > is captured by
+    # the generated Arr which contains the element's attributes
+    @openingElementLocationData = jisonLocationDataToAstLocationData @attributes.base.locationData
+
+    tagName = @tagName.base
+    tagName.locationData = tagName.tagNameLocationData
+    if @content?
+      @closingElementLocationData = mergeAstLocationData(
+        jisonLocationDataToAstLocationData tagName.closingTagOpeningBracketLocationData
+        jisonLocationDataToAstLocationData tagName.closingTagClosingBracketLocationData
+      )
+
+    super()
+
+  astType: ->
+    'JSXElement'
```

Comment:
If I were doing this "from scratch", I'd just name the class `JSXElement` and not have to override `astType()`

But I named it `CSXElement` since we seem to be using the naming convention `CSX` (as opposed to `JSX`)

I'm not convinced that's worth doing (it really _is_ JSX, so not sure what the benefit is of labeling it with a different name "CSX"), but not going to push to rename everything `CSX` -> `JSX`

Reply:
Yeah I’ve been planning on renaming all the CSX stuff to JSX for a long time now. The original author was fond of CSX as the name for “JSX in CoffeeScript” and hence the code was written that way and the PR for it described it as such, but when we wrote the docs for it we quickly realized that we had to call it JSX or else no one would have any idea what we were talking about; especially since it exactly matched the syntax for JSX. So anyway I’m 100% on board with just renaming all the CSX references to JSX if you want. We could maybe do it on a PR against `master` if that makes sense (and then merge that into this).

Reply:
Ah ok. My instinct is to do that in a separate commit/PR (after we're through the JSX AST stuff like you suggested) but do it against this branch (`ast`) rather than `master` - there will be enough updates to JSX-related code in this branch that I think it'd be more merging pain that it was worth to temporarily additionally target `master`